### PR TITLE
Add polygamma and zeta function to tensorflow

### PR DESCRIPTION
**Pull request by tillahoffmann**
This PR adds the polygamma and zeta functions as discussed in #1741. It also implements the derivative of the digamma function. Tests are provided and can be run provided `scipy` is installed.

### CODE REVIEW

**Comment by tillahoffmann** [comment]
I signed it!

File: tensorflow/core/ops/math_ops.cc [hunk 1]
```diff
@@ -433,6 +433,37 @@ Note, above `Q(a, x)` (`Igammac`) is the upper regularized complete
 Gamma function.
 )doc");
 
+REGISTER_OP("Zeta")
+    .Input("x: T")
+    .Input("q: T")
+    .Output("z: T")
+    .Attr("T: {float, double}")
+    .Doc(R"doc(
+Compute the Hurwitz zeta function `\zeta(x, q)`.
```
Comment (ebrevdo) [inline tensorflow/core/ops/math_ops.cc]: Strange code here?, \z.  Is this valid ascii?

File: tensorflow/core/ops/math_ops.cc [hunk 2]
```diff
@@ -433,6 +433,37 @@ Note, above `Q(a, x)` (`Igammac`) is the upper regularized complete
...
+
+The Hurwitz zeta function is defined as:
+
+```
+\zeta(x, q) = \sum_{n=0}^{\infty} (q + n)^{-x}
```
Comment (ebrevdo) [inline tensorflow/core/ops/math_ops.cc]: ditto here; and elsewhere.

**Comment by ebrevdo** [comment]
Overall looks good, though I am worried about the integer first argument.  It is mainly an issue because of the proto serialization and working across multiple machines; it's not always gonna be the case that if i create a float 2.0, and then it undergoes proto serialization, then gets passed to the op on a different architecture, that floor(n) == n anymore.

Would you consider making a functor that adds a small epsilon (e.g. 0.01) , then floors the integer input, just before calling the Eigen tensor method?

**Comment by gunan** [comment]
Can one of the admins verify this patch?

File: tensorflow/core/kernels/cwise_ops.h [hunk 3]
```diff
@@ -512,6 +512,12 @@ template <typename T>
 struct igammac : base<T, Eigen::internal::scalar_igammac_op<T>> {};
 
 template <typename T>
+struct zeta : base<T, Eigen::internal::scalar_zeta_op<T>> {};
```
Comment (ebrevdo) [inline tensorflow/core/kernels/cwise_ops.h]: I think this is where you'll need to extend the functors a bit.

File: tensorflow/core/ops/math_ops.cc (hunk 1 above)
Comment (tillahoffmann) [inline tensorflow/core/ops/math_ops.cc]: Just realised that you use \\( math \\) for inline math. Shall I fix all the other documentation characters in the process? E.g. see line [122](https://github.com/tillahoffmann/tensorflow/blob/master/tensorflow/core/ops/math_ops.cc#L122) in the same file.

**Comment by tillahoffmann** [comment]
What would be your opinion on demanding `abs(round(n) - n) < epsilon` in the eigen implementation or do you think this is TensorFlow specific because of the serialisation?

**Comment by ebrevdo** [comment]
i think the abs(round(n) - n) < epsilon, where epsilon is e.g. 10*machine
epsilon is a good solution.

**Comment by tillahoffmann** [comment]
I'm having second thoughts about the suggested implementation. I'm sure it works fine but I'm a bit worried about consistency: a number of functions in `SpecialFunctions.h` in eigen use equality tests. For example, `igamma` tests for equality with zero and I also test for equality with one in the implementation of the zeta function. Is there something special about the representation of one and zero in memory that we don't have to worry about serialisation issues? Would it be worthwhile adding a helper function `integer_equal_impl` for all such comparisons?

**Comment by ebrevdo** [comment]
Perhaps I jumped the gun.  Looks like float and double can represent
integers up to and including 2^{mantissa bits + 1}.  So there seems to be
no need to worry in this case.  Even with the vagaries of
serializing/deserializing, I don't believe that a  number with exact FP
representation will be modified.

**Comment by tillahoffmann** [comment]
Great, it seems we can just leave the implementations as they are. I have fixed the inline math. The problem with unescaped characters appears to be a [new feature for raw strings](https://solarianprogrammer.com/2011/10/16/cpp-11-raw-strings-literals-tutorial/) in C++11. E.g. the following will compile and print the expected string:

    const char* test = R"doc(This is the \zeta function.)doc";
    std::cout << test;

**Comment by ebrevdo** [comment]
I think we're ready to import this, but I'll be away for a while.  Assigning to @benoitsteiner for final merge.

**Comment by tillahoffmann** [comment]
Just wanted to follow up on this PR.

**Comment by ebrevdo** [comment]
@tensorflow-jenkins test this please.

**Comment by ebrevdo** [comment]
I was looking at the eigen implementation and saw that there may be a missing registration for the CUDA (float4 or double2) packet ops.  Can you revisit and make sure you added all the correct items?

**Comment by ebrevdo** [comment]
@tensorflow-jenkins : test this please.

**Comment by ebrevdo** [comment]
@martinwicke jenkins doesn't like me? :(

**Comment by ebrevdo** [comment]
@tillahoffmann ignore my comment about Eigen.  I re-checked and it looks good.  Waiting on jenkins to review.

File: .idea/encodings.xml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ebrevdo) [inline .idea/encodings.xml]: please remove the .idea directory.

File: build.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+git submodule update --init
```
Comment (ebrevdo) [inline build.sh]: please remove this

File: tensorflow/python/kernel_tests/.idea/misc.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/.idea/misc.xml]: please remove all .idea directories

File: tensorflow/workspace.bzl~ [hunk 7]
```diff
@@ -0,0 +1,92 @@
+# TensorFlow external dependencies that can be loaded in WORKSPACE files.
```
Comment (ebrevdo) [inline tensorflow/workspace.bzl~]: please remove all files ending with tilde

**Comment by tillahoffmann** [comment]
Sorry, they sneaked past my .gitignore. Should be all cleaned up now.

File: tensorflow/python/ops/math_grad.py [hunk 8]
```diff
@@ -295,8 +295,11 @@ def _LgammaGrad(op, grad):
 
 
 @ops.RegisterGradient("Digamma")
-def _DigammaGrad(op, grad):  # pylint: disable=unused-argument
-  raise NotImplementedError("grad(Digamma) == Polygamma(1) is not implemented")
+def _DigammaGrad(op, grad):
+  """Compute gradient of the digamma function with respect to its argument."""
+  x = op.inputs[0]
+  with ops.control_dependencies([grad.op]):
+    return grad * math_ops.polygamma(array_ops.ones(array_ops.shape(x), dtype=x.dtype), x)
```
Comment (ebrevdo) [inline tensorflow/python/ops/math_grad.py]: i *think* polygamma will distribute correctly if you just use constant(1, dtype=x.dtype)

Comment (tillahoffmann) [inline tensorflow/python/ops/math_grad.py]: Yes, `constant(1)` seems to work fine.

**Comment by ebrevdo** [comment]
I'll try one running tests more time.  It's possible you have to remerge and fix any merge conflicts

**Comment by ebrevdo** [comment]
@tensorflow-jenkins : test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please. And stop ignoring @ebrevdo.

**Pull request merged**